gridlab-d/gridlab-d
# MySQL Recorder and Group Recorder Improvements

#### What's this Pull Request do?
Supersedes Pull Request #1093.
#1060: Adds a MySQL implementation of the group_recorder class, as well as tests and support components needed to make the implementation viable. 
#1095: Adds improvements to the MySQL Recorder which allow for 'group' mode as well as making use of the query engine and buffering structures built for group_recorder. 
#1063: This bug was fixed as a necessary prerequisite for having the improvements made for #1095 work. 

#### Where should the reviewer start?
The reviewer will need to ensure gridlab-d is built and installed with MySQL support, and that the MySQL Connector C is installed and linked on their machine. 
To verify correct configuration and system compatibility, the "validate.no" file in the "mysql" directory should be removed, and gridlab-d's validation and test suite should be run. 

#### How should this be tested?
#1060: The system should be run using the MySQL module to replace the tape module for several group recorders of varying size, interval, and output count, to verify backwards compatibility with existing group recorders.
Additionally, the system should be tested with bespoke group recorders designed to test the various system functionalities and input options as listed on the [MySQL Group Recorder Wiki Page](http://gridlab-d.shoutwiki.com/wiki/Group_recorder_(mysql)).
Any test cases which might push the edge cases of the system should be run, and verified to either work, or fail in a predictable and resolvable way. 
#1095: The updated MySQL Recorder should be tested against several existing recorder examples to verify compatibility. 
#1063/#1095: New recorders should be assembled and tested making use of the new complex part attribute of the property parameter (see mysql/autotest/test_mysql_recorder_properties.glm for examples). 
#1095: New recorders should be assembled and tested making use of the new group parameter. The output should be verified to be data consistent with both the tape and mysql group recorders. 

#### What are the relevant issues?
#1060: MySQL has a hard row-size limit can cause the application to fail out during startup if the simulation produces more than 4096 columns or 65,535 bytes, whichever is smaller. This issue is generally addressed by the auto-set column_limit value, however it is possible to manually increase the value to levels where the system will fault.
Due to the overhead of initializing tables, and transferring data to MySQL, this implementation is slightly slower than the tape group recorder under most conditions, however if the simulation is small enough, and the query_buffer_limit is set low enough, the database interaction overhead can dominate the runtime, and significantly hinder use. 
#1095: There are some known risks surrounding the query_buffer_limit value and the user being able to modify it, as the property directly controls how much data is set to the database at a time. A value of 1 essentially acts in the same way as the old recorder (but with added overhead) and hinders performance. A large value (depends on the size of the MySQL cache configuration and the amount of data per query, but on the order of several thousand) in the query_buffer_limit can result in the buffered queries exceeding MySQL's cache length, resulting in MySQL halting the program until it has ingested the first section of data, before accepting the remaining data, which significantly affects performance. 

#### Questions:
No.
Yes. 

[validate.txt](https://github.com/gridlab-d/gridlab-d/files/1856933/validate.txt)
